filebeat/input/filestream/takeover/takeover_test.go

```diff
@@ -0,0 +1,313 @@
+				"source":    "/path/NOMATCH.log",
+				"timestamp": []int{258139663760, 1671033742},
+				"ttl":       -1,
+				"id":        "native::33333333-44444444",
+				"offset":    1,
+				"type":      "log",
+				"FileStateOS": mapstr.M{
+					"inode":  33333333,
+					"device": 44444444,
+				},
+				"identifier_name": "native",
+				"prev_id":         "",
+			},
+		},
+	}
+
+	cases := []struct {
+		name       string
+		cfg        *conf.C
+		states     []state
+		mustBackup bool
+		mustRemove []string
+		mustSet    []setOp
+		expErr     string
+	}{
+		{
+			name:   "does nothing when default config",
+			cfg:    empty,
+			states: states,
+		},
+		{
+			name:   "does nothing when there is no filestream with `take_over`",
+			cfg:    noTakeOver,
+			states: states,
+		},
+		{
+			name:       "filestream takes over when there is `take_over: true`",
+			cfg:        takeOver,
+			states:     states,
+			mustBackup: true,
+			mustRemove: []string{
+				states[1].key,
+				states[2].key,
+			},
+			mustSet: []setOp{
+				{
+					key: "filestream::filestream-id-2::native::92938222-16777232",
+					value: mapstr.M{
+						"meta": mapstr.M{
+							"source":          "/path/log1.log",
+							"identifier_name": "native",
+						},
+						"ttl":     -1,
+						"updated": []int{258139663760, 1671033742},
+						"cursor": mapstr.M{
+							"offset": 392012100,
+						},
+					},
+				},
+				{
+					key: "filestream::filestream-id-2::native::92938223-16777233",
+					value: mapstr.M{
+						"meta": mapstr.M{
+							"source":          "/path/log2.log",
+							"identifier_name": "native",
+						},
+						"ttl":     -1,
+						"updated": []int{258139663761, 1671033743},
+						"cursor": mapstr.M{
+							"offset": 64625356,
+						},
+					},
+				},
+			},
+		},
+		{
+			name:   "does nothing when there is no matching loginput entry",
+			cfg:    takeOver,
+			states: states[3:],
+		},
+	}
+
+	log := logp.NewLogger("takeover-test")
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			config := cfg.DefaultConfig
+			err := tc.cfg.Unpack(&config)
+			require.NoError(t, err)
+
+			store := storeMock{
+				states: tc.states,
+			}
+			backuper := backuperMock{}
+			err = TakeOverLogInputStates(log, &store, &backuper, &config)
+			if tc.expErr != "" {
+				require.Error(t, err)
+				require.Contains(t, err.Error(), tc.expErr)
+				return
+			}
+			require.NoError(t, err)
+
+			if tc.mustBackup {
+				require.Equal(t, 1, backuper.called, "backup must be called exactly once")
+			} else {
+				require.Equal(t, 0, backuper.called, "backup must not be called")
+			}
+
+			sort.Strings(tc.mustRemove)
+			sort.Slice(tc.mustSet, func(i, j int) bool {
+				return tc.mustSet[i].key < tc.mustSet[j].key
+			})
+			sort.Strings(store.removed)
+			sort.Slice(store.set, func(i, j int) bool {
+				return store.set[i].key < store.set[j].key
+			})
+
+			require.Equal(t, tc.mustRemove, store.removed)
+			require.Equal(t, tc.mustSet, store.set)
```

Comment:
[Suggestion]

Use [`require.ElementsMatch`](https://pkg.go.dev/github.com/stretchr/testify/require#ElementsMatch) so you don't need to sort them manually.

```suggestion
			require.ElementsMatch(t, tc.mustRemove, store.removed)
			require.ElementsMatch(t, tc.mustSet, store.set)
```

Reply:
Didn't know that function existed 😀